iml-services/iml-snapshot/src/retention.rs

```diff
@@ -0,0 +1,215 @@
+        if nodes.is_empty() {
+            return Ok(None);
+        } else {
+            let bytes_total = nodes[0].bytes_total;
+            let bytes_free = nodes[0].bytes_free;
+            let bytes_used = bytes_total - bytes_free;
+
+            return Ok(Some((bytes_total, bytes_free, bytes_used)));
+        }
+    }
+
+    Ok(None)
+}
+
+async fn get_snapshots(
+    pool: &sqlx::PgPool,
+    fs_name: &str,
+) -> Result<Vec<snapshot::SnapshotRecord>, Error> {
+    let xs = sqlx::query_as!(
+        snapshot::SnapshotRecord,
+        "SELECT * FROM snapshot WHERE filesystem_name = $1 ORDER BY create_time ASC",
+        fs_name
+    )
+    .fetch_all(pool)
+    .await?;
+
+    Ok(xs)
+}
+
+async fn get_retentions(pool: &sqlx::PgPool) -> Result<Vec<snapshot::SnapshotRetention>, Error> {
+    let xs = sqlx::query_as!(
+        snapshot::SnapshotRetention,
+        r#"
+                SELECT
+                    id,
+                    filesystem_name,
+                    reserve_value,
+                    reserve_unit as "reserve_unit:snapshot::ReserveUnit",
+                    last_run,
+                    keep_num
+                FROM snapshot_retention
+            "#
+    )
+    .fetch_all(pool)
+    .await?;
+
+    Ok(xs)
+}
+
+async fn get_retention_policy(
+    pool: &PgPool,
+    fs_name: &str,
+) -> Result<Option<snapshot::SnapshotRetention>, Error> {
+    let xs = get_retentions(pool).await?;
+
+    Ok(xs.into_iter().find(|x| x.filesystem_name == fs_name))
+}
+
+async fn destroy_snapshot(
+    client: Client,
+    fs_name: &str,
+    snapshot_name: &str,
+) -> Result<Command, Error> {
+    let resp: iml_graphql_queries::Response<iml_graphql_queries::snapshot::destroy::Resp> =
+        graphql(
+            client,
+            iml_graphql_queries::snapshot::destroy::build(fs_name, snapshot_name, true),
+        )
+        .await?;
+
+    let cmd = Result::from(resp)?.data.destroy_snapshot;
+
+    Ok(cmd)
+}
+
+async fn get_retention_filesystems(pool: &PgPool) -> Result<Vec<String>, Error> {
+    let xs = get_retentions(pool).await?;
+
+    let fs_names = xs.into_iter().map(|x| x.filesystem_name).collect();
+    Ok(fs_names)
+}
+
+pub async fn process_retention(
+    client: &Client,
+    influx_client: &InfluxClient,
+    pool: &PgPool,
+    mut stats_record: HashMap<String, u64>,
+) -> Result<HashMap<String, u64>, Error> {
+    let filesystems = get_retention_filesystems(pool).await?;
+
+    tracing::error!("Filesystems with retentions: {:?}", filesystems);
+
+    for fs_name in filesystems {
+        let stats = get_stats_from_influx(&fs_name, &influx_client).await?;
+
+        if let Some((bytes_total, bytes_free, bytes_used)) = stats {
+            tracing::debug!(
+                "stats values: {}, {}, {}",
+                bytes_total,
+                bytes_free,
+                bytes_used
+            );
+            let percent_used = bytes_used as f64 / bytes_total as f64 * 100.0f64;
```

Comment:
```suggestion
            let percent_used = (bytes_used as f64 / bytes_total as f64) * 100.0f64;
```

nit: I know the priorities, but like them explicit :)

Reply:
Sure, I'll add parenthesis.